POEM_056.md

```diff
@@ -0,0 +1,349 @@
+POEM ID: 056  
+Title:  Function based API usable by OpenMDAO and others  
+authors: [Bret Naylor]  
+Competing POEMs: 052    
+Related POEMs: 039   
+Associated implementation PR:  
+
+##  Status
+
+- [x] Active
+- [ ] Requesting decision
+- [ ] Accepted
+- [ ] Rejected
+- [ ] Integrated
+
+
+First a note about competing POEM 52.  This POEM is more limited in scope than POEM 52 and 
+focuses only on the API for associating metadata with and retrieving metadata from a function
+object.  POEM 52 also deals with the OpenMDAO function component(s) that will use the API it defines.
+This POEM does not describe those components in any detail and will defer that discussion to a later 
+POEM.
+
+
+## Motivation
+
+Define an API for attaching metadata to and retrieving metadata from a function object.  This API
+will enable a user to attach metadata to a function object sufficient for OpenMDAO to create a 
+fully operational Component, including input and output variable names, shapes and units as well
+as partial derivative and coloring information.  The body of the function will act as the 
+component's `compute` method.
+
+Depending upon the importance of eliminating any dependence on OpenMDAO, this API could either be
+released as a standalone distribution called, for example, `openmdao_func_api`, or as just a
+sub-package within the openmdao distribution, e.g., `openmdao.func_api`.
+
+For the rest of this document, assume that we've imported this package as either
+
+```python
+import openmdao_func_api.api as omf
+```
+
+or
+
+```python
+import openmdao.func_api as omf
+```
+
+A two-way API as described above, where API functions are used to both attach and retrieve data,
+allows the underlying data layout and location to be hidden and potentially updated later without 
+negatively impacting users of the API.
+
+This API consists primarily of methods on the OMWrappedFunc class, which wraps a plain python or
+potentially annotated function and stores metadata needed by openmdao along with that function.
```

Comment:
You say "potentially annotated" but don't describe what you mean by that. 

Are you saying that you could accept the I/O annotations laid out in POEM 052? If so, this POEM should stand alone, so please pull the relevant details into here and give at least one example. 

Im not sure what the annotations really buy you here though. If we track the order of the `add_input` and `add_output` calls then you wouldn't even need to provide the arg names (at least for inputs and cases where the output names could be inspected). 

I think the annotations are not needed, and we'd be better of giving examples of how we could use the AST to iner the names instead (and explain the case when return values are expressions when inference isn't possible and names will be required).

Reply:
I dropped the 'potentially annotated' stuff and added a little more explanation of output name inference.